lib/spack/spack/test/concretize.py

```diff
@@ -190,6 +191,28 @@ def test_different_compilers_get_different_flags(self):
         assert set(client.compiler_flags['fflags']) == set(['-O0'])
         assert not set(cmake.compiler_flags['fflags'])
 
+    def test_architecture_inheritance(self):
+        spec = Spec('cmake-client %gcc@4.7.2 os=fe ^ cmake')
+        try:
+            spec.concretize()
+        except UnavailableCompilerVersionError:
+            # Concretization will fail with an UnavailableCompilerVersionError
+            # if the architecture is concretized incorrectly.
+            assert False
```

Comment:
@becker33 why not just let the exception through? In that case the test will fail also and with less logic

Reply:
@scheibelp I could have it pass and fall through and fail on the later test, but I already know it will fail at this point, it seems cleaner to me to just fail it when I know it has failed.

Reply:
The exception would cause the test to fail on this line (or rather on the `spec.concretize()` line). At least I'm almost sure that it should

Reply:
@scheibelp If I didn't catch the exception at all it would fail with an exception rather than with a failure. I would rather produce a failure for an exception that I expect than an exception.

Reply:
@becker33 this doesn't appear to preserve the stack trace for the series of calls that led to the `UnavailableCompilerError` (i.e. it will not include the tree of calls under `spec.concretize`). I'm curious about the advantage of reporting a unit test that would otherwise terminate with an exception as a failure. I'm assuming the intent is to provide the user with an indication of where things could have gone wrong, as indicated by the comment:

> Concretization will fail with an UnavailableCompilerVersionError if the architecture is concretized incorrectly.

but my points would be that

* there are other reasons that the error could come up (e.g. if the test `compilers.yaml` is edited and an error is introduced there)
* IMO the comment adds all of the useful information - without the comment, catching the exception doesn't give the user any hint (since the exception would have been reported by pytest whether or not it was caught)

Reply:
@becker33: I think the right thing to do if you must use your own try/catch is to catch and use `pytest.fail("blah")` so that you actually spit the cause of the error out to the console.

Though IMHO, I agree with Peter -- it's cleaner to just run the code and just let the exception through.  You can add context to that by adding a docstring for the test function -- the docstrings are shown in pytest output.